common/elasticsearch/bulk/bulk_index_request.go

```diff
@@ -0,0 +1,253 @@
+// for details.
+func (r *BulkIndexRequest) VersionType(versionType string) *BulkIndexRequest {
+	r.versionType = versionType
+	r.source = nil
+	return r
+}
+
+// Doc specifies the document to index.
+func (r *BulkIndexRequest) Doc(doc interface{}) *BulkIndexRequest {
+	r.doc = doc
+	r.source = nil
+	return r
+}
+
+// RetryOnConflict specifies how often to retry in case of a version conflict.
+func (r *BulkIndexRequest) RetryOnConflict(retryOnConflict int) *BulkIndexRequest {
+	r.retryOnConflict = &retryOnConflict
+	r.source = nil
+	return r
+}
+
+// Pipeline to use while processing the request.
+func (r *BulkIndexRequest) Pipeline(pipeline string) *BulkIndexRequest {
+	r.pipeline = pipeline
+	r.source = nil
+	return r
+}
+
+// IfSeqNo indicates to only perform the index operation if the last
+// operation that has changed the document has the specified sequence number.
+func (r *BulkIndexRequest) IfSeqNo(ifSeqNo int64) *BulkIndexRequest {
+	r.ifSeqNo = &ifSeqNo
+	return r
+}
+
+// IfPrimaryTerm indicates to only perform the index operation if the
+// last operation that has changed the document has the specified primary term.
+func (r *BulkIndexRequest) IfPrimaryTerm(ifPrimaryTerm int64) *BulkIndexRequest {
+	r.ifPrimaryTerm = &ifPrimaryTerm
+	return r
+}
+
+// String returns the on-wire representation of the index request,
+// concatenated as a single string.
+func (r *BulkIndexRequest) String() string {
+	lines, err := r.Source()
+	if err != nil {
+		return fmt.Sprintf("error: %v", err)
+	}
+	return strings.Join(lines, "\n")
+}
+
+// Source returns the on-wire representation of the index request,
+// split into an action-and-meta-data line and an (optional) source line.
+// See https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docs-bulk.html
+// for details.
+func (r *BulkIndexRequest) Source() ([]string, error) {
+	// { "index" : { "_index" : "test", "_type" : "type1", "_id" : "1" } }
+	// { "field1" : "value1" }
+
+	if r.source != nil {
+		return r.source, nil
+	}
+
+	lines := make([]string, 2)
+
+	// "index" ...
+	indexCommand := bulkIndexRequestCommandOp{
+		Index:           r.index,
+		Type:            r.typ,
+		ID:              r.id,
+		Routing:         r.routing,
+		Parent:          r.parent,
+		Version:         r.version,
+		VersionType:     r.versionType,
+		RetryOnConflict: r.retryOnConflict,
+		Pipeline:        r.pipeline,
+		IfSeqNo:         r.ifSeqNo,
+		IfPrimaryTerm:   r.ifPrimaryTerm,
+	}
+	command := bulkIndexRequestCommand{
+		r.opType: indexCommand,
+	}
+
+	var err error
+	var body []byte
+	body, err = json.Marshal(command)
+
+	if err != nil {
+		return nil, err
+	}
+
+	lines[0] = string(body)
+
+	// "field1" ...
+	if r.doc != nil {
+		switch t := r.doc.(type) {
+		default:
+			body, err := json.Marshal(r.doc)
+			if err != nil {
+				return nil, err
+			}
+			lines[1] = string(body)
+		case json.RawMessage:
+			lines[1] = string(t)
+		case *json.RawMessage:
+			lines[1] = string(*t)
```

Comment:
Wouldn't this be handled by `json.Marshal` too, for "free"?

Reply:
yea, these are handled by json.Marshal:
```
func TestJSON(t *testing.T) {
	orig := `{"valid":"json"}`
	t.Run("raw value", func(t *testing.T) {
		r := json.RawMessage(orig)
		s, err := json.Marshal(r)
		require.NoError(t, err)
		require.Equal(t, string(s), orig)
	})
	t.Run("raw pointer", func(t *testing.T) {
		r := json.RawMessage(orig)
		s, err := json.Marshal(&r)
		require.NoError(t, err)
		require.Equal(t, string(s), orig)
	})
}
```
it just passes through, essentially "for free" (but safer, because it does still validate that it's valid json, rather than throwing it in regardless)

Reply:
done